cuda/matrix/csr_kernels.cu

```diff
@@ -634,20 +634,86 @@ void advanced_spmv(std::shared_ptr<const CudaExecutor> exec,
 GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
     GKO_DECLARE_CSR_ADVANCED_SPMV_KERNEL);
 
+namespace kernel
+{
+
+template <typename IndexType>
+__global__ __launch_bounds__(default_block_size) void convert_row_ptrs_to_idxs(size_type num_rows,
+		const IndexType* ptrs, IndexType* idxs)
+{
+	const auto tidx = threadIdx.x + blockDim.x * blockIdx.x;
+	if (tidx < num_rows) {
+		for (auto i = ptrs[tidx]; i < ptrs[tidx + 1]; i++) {
+			idxs[i] = tidx;
+		}
+	}
+}
+
+} // namespace kernel
 
 template <typename IndexType>
 void convert_row_ptrs_to_idxs(std::shared_ptr<const CudaExecutor> exec,
                               const IndexType *ptrs, size_type num_rows,
-                              IndexType *idxs) NOT_IMPLEMENTED;
+                              IndexType *idxs)
+{
+	auto grid_dim = ceildiv(num_rows, default_block_size);
+
+	kernel::convert_row_ptrs_to_idxs<<<grid_dim, default_block_size>>>(num_rows, as_cuda_type(ptrs),
+			as_cuda_type(idxs));
+}
 
 GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(
     GKO_DECLARE_CSR_CONVERT_ROW_PTRS_TO_IDXS_KERNEL);
 
+namespace kernel
+{
+
+template <typename ValueType>
+__global__ void initialize_zero_dense(size_type num_rows, size_type num_cols,
+		size_type stride, ValueType* result)
+{
+	auto tidx_x = threadIdx.x + blockDim.x * blockIdx.x;
+	auto tidx_y = threadIdx.y + blockDim.y * blockIdx.y;
+	if (tidx_x < num_rows && tidx_y < num_cols) {
+		result[tidx_x * stride + tidx_y] = zero<ValueType>();
+	}
+}
+
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(default_block_size) void fill_in_dense(size_type num_rows, const IndexType* row_ptrs,
+		const IndexType* col_idxs, const ValueType* values, size_type stride, ValueType* result)
+{
+	const auto tidx = threadIdx.x + blockDim.x * blockIdx.x;
+	if (tidx < num_rows) {
+		for (auto i = row_ptrs[tidx]; i < row_ptrs[tidx + 1]; i++) {
+			result[stride * tidx + col_idxs[i]] = values[i];
+		}
+	}
+}
+
+} // namespace kernel
 
 template <typename ValueType, typename IndexType>
 void convert_to_dense(
     std::shared_ptr<const CudaExecutor> exec, matrix::Dense<ValueType> *result,
-    const matrix::Csr<ValueType, IndexType> *source) NOT_IMPLEMENTED;
+    const matrix::Csr<ValueType, IndexType> *source)
+{
+	auto num_rows = result->get_size()[0];
+	auto num_cols = result->get_size()[1];
+	auto stride = result->get_stride();
+	auto row_ptrs = source->get_const_row_ptrs();
+	auto col_idxs = source->get_const_col_idxs();
+	auto vals = source->get_const_values();
+
+	dim3 block_size = (32, 32, 1);
+	dim3 init_grid_dim = (ceildiv(num_rows, 32), ceildiv(num_cols, 32), 1);
```

Comment:
All variables in this function should be `const`.
Also remove the `=` while calling the `dim3` constructor:
```c++
dim3 block_size(32, 32, 1);
dim3 init_grid_dim(ceildiv(num_rows, 32), ceildiv(num_cols, 32), 1);
```